### PR TITLE
Add ability to append lifecycle and repository policies for ECR

### PR DESCRIPTION
**Pull request by mhumeSF**
This'll allow one to specify a json file in the repo that has a lifecycle or repository policy that'll be applied to the ECR repo.

### CODE REVIEW

**Comment by mhumeSF** [comment]
This'll allow one to specify a json file in the repo that has a lifecycle or repository policy that'll be applied to the ECR repo.

File: cmd/drone-ecr/main.go [hunk 1]
```diff
@@ -19,11 +20,13 @@ const defaultRegion = "us-east-1"
...
-		region = getenv("PLUGIN_REGION", "ECR_REGION", "AWS_REGION")
-		key    = getenv("PLUGIN_ACCESS_KEY", "ECR_ACCESS_KEY", "AWS_ACCESS_KEY_ID")
-		secret = getenv("PLUGIN_SECRET_KEY", "ECR_SECRET_KEY", "AWS_SECRET_ACCESS_KEY")
-		create = parseBoolOrDefault(false, getenv("PLUGIN_CREATE_REPOSITORY", "ECR_CREATE_REPOSITORY"))
+		repo              = getenv("PLUGIN_REPO")
+		region            = getenv("PLUGIN_REGION", "ECR_REGION", "AWS_REGION")
+		key               = getenv("PLUGIN_ACCESS_KEY", "ECR_ACCESS_KEY", "AWS_ACCESS_KEY_ID")
+		secret            = getenv("PLUGIN_SECRET_KEY", "ECR_SECRET_KEY", "AWS_SECRET_ACCESS_KEY")
+		create            = parseBoolOrDefault(false, getenv("PLUGIN_CREATE_REPOSITORY", "ECR_CREATE_REPOSITORY"))
+		lifecycle_policy  = getenv("PLUGIN_LIFECYCLE_POLICY")
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: Please use camelcase, e.g. lifecyclePolicy

File: cmd/drone-ecr/main.go [hunk 2]
```diff
@@ -19,11 +20,13 @@ const defaultRegion = "us-east-1"
...
+		repository_policy = getenv("PLUGIN_REPOSITORY_POLICY")
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: Same here

File: cmd/drone-ecr/main.go [hunk 3]
```diff
@@ -95,6 +122,30 @@ func ensureRepoExists(svc *ecr.ECR, name string) (err error) {
 	return
 }
 
+func uploadLifeCyclePolicy(svc *ecr.ECR, lifecycle_policy string, name string) (err error) {
+	input := &ecr.PutLifecyclePolicyInput{}
+	input.SetLifecyclePolicyText(lifecycle_policy)
+	input.SetRepositoryName(name)
+	_, err = svc.PutLifecyclePolicy(input)
+	if err != nil {
+        log.Fatal(fmt.Sprintf("error uploading ECR lifecycle policy. %v", err))
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: Wrong indentation

File: cmd/drone-ecr/main.go [hunk 4]
```diff
@@ -95,6 +122,30 @@ func ensureRepoExists(svc *ecr.ECR, name string) (err error) {
...
+	return
+}
+
+func uploadRepositoryPolicy(svc *ecr.ECR, repository_policy string, name string) (err error) {
+	input := &ecr.SetRepositoryPolicyInput{}
+	input.SetPolicyText(repository_policy)
+	input.SetRepositoryName(name)
+	_, err = svc.SetRepositoryPolicy(input)
+	if err != nil {
+        log.Fatal(fmt.Sprintf("error uploading ECR repository policy. %v", err))
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: Same here

File: cmd/drone-ecr/main.go [hunk 5]
```diff
@@ -95,6 +122,30 @@ func ensureRepoExists(svc *ecr.ECR, name string) (err error) {
 	return
 }
 
+func uploadLifeCyclePolicy(svc *ecr.ECR, lifecycle_policy string, name string) (err error) {
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: Please don't use named returns

File: cmd/drone-ecr/main.go [hunk 6]
```diff
@@ -95,6 +122,30 @@ func ensureRepoExists(svc *ecr.ECR, name string) (err error) {
...
+	input.SetRepositoryName(name)
+	_, err = svc.PutLifecyclePolicy(input)
+	if err != nil {
+        log.Fatal(fmt.Sprintf("error uploading ECR lifecycle policy. %v", err))
+	}
+
+	return
+}
+
+func uploadRepositoryPolicy(svc *ecr.ECR, repository_policy string, name string) (err error) {
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: Same here

File: cmd/drone-ecr/main.go (hunk 3 above)
Comment (mhumeSF) [reply cmd/drone-ecr/main.go]: Updated to camelcase

File: cmd/drone-ecr/main.go (hunk 2 above)
Comment (mhumeSF) [reply cmd/drone-ecr/main.go]: Updated to camelcase

File: cmd/drone-ecr/main.go (hunk 1 above)
Comment (mhumeSF) [reply cmd/drone-ecr/main.go]: Updated to camelcase

File: cmd/drone-ecr/main.go (hunk 4 above)
Comment (mhumeSF) [reply cmd/drone-ecr/main.go]: Corrected with `go fmt ...`

File: cmd/drone-ecr/main.go (hunk 5 above)
Comment (mhumeSF) [reply cmd/drone-ecr/main.go]: Sure thing. Updated to `return err`

**Comment by mhumeSF** [comment]
If that looks ok, I'll rebase to single commit before merge

File: cmd/drone-ecr/main.go [hunk 7]
```diff
@@ -95,6 +122,30 @@ func ensureRepoExists(svc *ecr.ECR, name string) (err error) {
 	return
 }
 
+func uploadLifeCyclePolicy(svc *ecr.ECR, lifecyclePolicy string, name string) (err error) {
+	input := &ecr.PutLifecyclePolicyInput{}
+	input.SetLifecyclePolicyText(lifecyclePolicy)
+	input.SetRepositoryName(name)
+	_, err = svc.PutLifecyclePolicy(input)
+	if err != nil {
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: You can skip that and always return err

File: cmd/drone-ecr/main.go [hunk 8]
```diff
@@ -95,6 +122,30 @@ func ensureRepoExists(svc *ecr.ECR, name string) (err error) {
...
+
+	return err
+}
+
+func uploadRepositoryPolicy(svc *ecr.ECR, repositoryPolicy string, name string) (err error) {
+	input := &ecr.SetRepositoryPolicyInput{}
+	input.SetPolicyText(repositoryPolicy)
+	input.SetRepositoryName(name)
+	_, err = svc.SetRepositoryPolicy(input)
+	if err != nil {
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: You can skip that and always return err

File: cmd/drone-ecr/main.go [hunk 9]
```diff
@@ -61,6 +64,30 @@ func main() {
 		}
 	}
 
+	if lifecyclePolicy != "" {
+		p, err := ioutil.ReadFile(lifecyclePolicy)
+		if err != nil {
+			log.Fatal(fmt.Sprintf("error reading ECR lifecycle policy: %v", err))
+		}
+
+		err = uploadLifeCyclePolicy(svc, string(p), trimHostname(repo, registry))
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: Please add this into the if condition to have a single line.

File: cmd/drone-ecr/main.go [hunk 10]
```diff
@@ -61,6 +64,30 @@ func main() {
...
+		}
+	}
+
+	if repositoryPolicy != "" {
+        p, err := ioutil.ReadFile(repositoryPolicy)
+		if err != nil {
+			log.Fatal(fmt.Sprintf("error reading ECR repository policy. %v", err))
+		}
+
+		err = uploadRepositoryPolicy(svc, string(p), trimHostname(repo, registry))
```
Comment (tboerger) [inline cmd/drone-ecr/main.go]: Please add this into the if condition to have a single line.

File: cmd/drone-ecr/main.go (hunk 8 above)
Comment (mhumeSF) [reply cmd/drone-ecr/main.go]: refactored

File: cmd/drone-ecr/main.go (hunk 7 above)
Comment (mhumeSF) [reply cmd/drone-ecr/main.go]: Ah, this makes much more sense to have handled with the if statement

**Comment by mhumeSF** [comment]
@bradrydzewski This cool to merge?

**Comment by mhumeSF** [comment]
@tboerger @appleboy Thanks for review 👍

**Pull request merged**